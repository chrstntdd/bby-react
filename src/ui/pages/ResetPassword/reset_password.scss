@import '$homeDir/styles/rwd-helpers.scss';
@import '$homeDir/styles/colors.scss';

#reset-password-wrapper {
  min-height: 100vh;
  width: 100vw;
  #reset-password-card {
    margin-top: 6rem;
    border-top: 0.3rem solid $BESTBUY_BLUE;
    padding: 0;
    width: between(300px, 660px, $bp-xxsmall, $bp-xlarge);
    padding: between(0, 16px, $bp-xxsmall, $bp-xlarge);
    form {
      width: between(300px, 560px, $bp-xxsmall, $bp-xlarge);
      margin: 0 auto;
    }
  }
}
