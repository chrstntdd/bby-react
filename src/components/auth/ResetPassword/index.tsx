import * as React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Field, reduxForm } from 'redux-form';
import { resetPassword } from '../../../state/actions';
import LoadingIndicator from '../Loading';
import 'url-search-params-polyfill';

import Input from '../FormInput';
import './form-input.scss';

import './reset_password.scss';

export const validate = formProps => {
  const errors = {};

  if (!formProps.password) {
    errors.password = 'Please enter a new password';
  }

  if (!formProps.passwordConfirm) {
    errors.passwordConfirm = 'Please confirm new password';
  }

  if (formProps.password !== formProps.passwordConfirm) {
    errors.password = 'Passwords must match';
  }

  return errors;
};

const form = reduxForm({
  form: 'resetPassword',
  validate
});

export class ResetPassword extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      inputs: [
        {
          name: 'password',
          type: 'password',
          label: 'Password'
        },
        {
          name: 'passwordConfirm',
          type: 'password',
          label: 'Confirm Password'
        }
      ]
    };
  }

  async handleFormSubmit({ password }, formProps) {
    const search = this.props.location.search;
    const params = new URLSearchParams(search);
    const resetToken = params.get('token');
    await this.props.resetPassword(resetToken, { password });
    this.props.history.push('/dashboard');
  }
  renderAlert() {
    if (this.props.errorMessage) {
      return (
        <div className="error-message">
          <p>{this.props.errorMessage}</p>
        </div>
      );
    } else if (this.props.message) {
      return (
        <div className="success-message">
          <p>{this.props.message}</p>
        </div>
      );
    }
  }
  render() {
    const { handleSubmit, waiting } = this.props;

    const formInputs = this.state.inputs.map((input, index) => (
      <Input key={index} {...input} />
    ));

    return (
      <section id="reset-password-wrapper">
        <LoadingIndicator waiting={waiting} message={'One moment please.'} />
        <div id="reset-password-card" className={waiting ? 'hide' : 'show'}>
          <h1>Reset Password</h1>
          <form onSubmit={handleSubmit(this.handleFormSubmit.bind(this))}>
            {this.renderAlert()}
            {formInputs}
            <button action="submit">Change Password</button>
          </form>
        </div>
      </section>
    );
  }
}

const mapStateToProps = state => ({
  waiting: state.auth.waiting,
  errorMessage: state.auth.error,
  message: state.auth.message
});

export default withRouter(
  connect(mapStateToProps, { resetPassword })(form(ResetPassword))
);
