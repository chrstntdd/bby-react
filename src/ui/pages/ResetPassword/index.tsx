import React, { Component, createRef } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { RouteComponentProps } from 'react-router';

import { validateInput, getFirstQueryParamVal } from '@/util';

import { resetPassword } from '@/state/routines';

import LoadingIndicator from '@/ui/components/Loading';
import Input from '@/ui/components/Input';
import Form from '@/ui/components/Form';

import './reset_password.scss';

interface PResetPassword {
  resetPassword: (resetToken: string, newPassword: string) => (dispatch: any) => Promise<void>;
  waiting: boolean;
  errorMessage: string;
  message: string;
}

interface SResetPassword {}

export class ResetPassword extends Component<
  PResetPassword & RouteComponentProps<{}>,
  SResetPassword
> {
  state = {
    fieldDefaults: [
      [
        'newPasswordInput',
        {
          validationFn: () =>
            validateInput(
              'Please enter a password with at least 6 characters',
              new RegExp(/.{6,}/, 'gi')
            )
        }
      ],
      [
        'confirmNewPasswordInput',
        {
          validationFn: () =>
            validateInput(
              'Please enter a password with at least 6 characters',
              new RegExp(/.{6,}/, 'gi')
            )
        }
      ]
    ]
  };

  private newPasswordInput: React.RefObject<HTMLInputElement> = createRef();
  private confirmNewPasswordInput: React.RefObject<HTMLInputElement> = createRef();

  handleFormSubmit = (formValues): void => {
    const [newPasswordInput, confirmNewPasswordInput] = formValues;
    const resetToken = getFirstQueryParamVal();

    if (newPasswordInput[1] !== confirmNewPasswordInput[1]) {
      /* TODO: handle this case */
      console.log('PASSWORD MISMATCH');
    } else {
      this.props.resetPassword(resetToken, newPasswordInput);
      this.props.history.push('/dashboard');
    }
  };

  renderAPIMsg() {
    if (this.props.errorMessage) {
      return (
        <div className="error-message">
          <p>{this.props.errorMessage}</p>
        </div>
      );
    } else if (this.props.message) {
      return (
        <div className="success-message">
          <p>{this.props.message}</p>
        </div>
      );
    } else {
      return null;
    }
  }

  legendClass = 'text-center text-2xl blue-accent';
  cardClass = 'bg-white-darkest mx-auto shadow-lg rounded-b overflow-hidden flex flex-col items-center justify-center';
  submitButtonClass = 'mx-auto my-2 font-semibold rounded-full px-8 py-2 leading-normal bg-transparent border border-grey text-grey hover:border-bby-blue hover:bg-bby-blue hover:text-white trans-300ms-all';
  fieldsetClass = 'px-2';

  render() {
    const { waiting } = this.props;

    return (
      <main id="reset-password-wrapper">
        {!waiting ? null : <LoadingIndicator waiting={waiting} message={'One moment please.'} />}

        {this.renderAPIMsg()}

        <div id="reset-password-card" className={waiting ? 'hide' : `show ${this.cardClass}`}>
          <Form
            id="reset-password"
            fieldDefaults={this.state.fieldDefaults}
            onFormSubmit={this.handleFormSubmit}
            className="flex flex-col"
            render={({ onChange, getInputProps, getFormState, getFormProps }) => {
              return (
                <React.Fragment>
                  <fieldset className={this.fieldsetClass}>
                    <legend id="reset-password" className={this.legendClass}>
                      Reset your password
                    </legend>

                    <div className="input-wrapper">
                      <Input
                        {...getInputProps({
                          type: 'password',
                          label: 'New Password',
                          id: 'newPasswordInput'
                        })}
                      />
                    </div>

                    <div className="input-wrapper">
                      <Input
                        {...getInputProps({
                          type: 'password',
                          label: 'Confirm New Password',
                          id: 'confirmNewPasswordInput'
                        })}
                      />
                    </div>
                  </fieldset>
                  <button
                    type="submit"
                    disabled={!getFormState().allFieldsValid}
                    className={this.submitButtonClass}
                  >
                    CHANGE PASSWORD AND LOGIN
                  </button>
                </React.Fragment>
              );
            }}
          />
        </div>
      </main>
    );
  }
}

const mapStateToProps = state => ({
  waiting: state.auth.waiting,
  errorMessage: state.auth.error,
  message: state.auth.message
});

export default withRouter(connect(mapStateToProps, { resetPassword })(ResetPassword));
