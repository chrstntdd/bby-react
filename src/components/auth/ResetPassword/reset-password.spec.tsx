import React from 'react';
import { shallow } from 'enzyme';

import ResetPassword from './';

describe('<ResetPassword/>', () => {
  it('should render without crashing', () => {
    const wrapper = shallow(<ResetPassword />);

    expect(wrapper).toMatchSnapshot();
  });
});
