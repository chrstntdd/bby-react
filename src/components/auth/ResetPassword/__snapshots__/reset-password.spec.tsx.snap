// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetPassword/> should render without crashing 1`] = `
<Route
  render={[Function]}
/>
`;
